### PR TITLE
[search] Removed Ranker's dependency on Processor.

### PR DESCRIPTION
**Pull request by mpimenov**

### CODE REVIEW

File: search/pre_ranker.hpp [hunk 1]
```diff
@@ -45,6 +46,8 @@ class PreRanker
       fn(result.GetId(), result.GetInfo());
   }
 
+  Ranker * m_ranker;
```
Comment (ygorshenin) [inline search/pre_ranker.hpp]: Raw pointer in a public section? Do we really need it?

File: search/processor.cpp [hunk 2]
```diff
@@ -376,59 +335,51 @@ int Processor::GetCategoryLocales(int8_t(&arr)[3]) const
...
+  Geocoder::Params geocoderParams;
+  InitParams(geocoderParams);
+  geocoderParams.m_mode = m_mode;
+  geocoderParams.m_pivot = GetPivotRect();
+  geocoderParams.m_accuratePivotCenter = GetPivotPoint();
+  m_geocoder.SetParams(geocoderParams);
+
+  Ranker::Params rankerParams;
+  rankerParams.m_currentLocaleCode = m_currentLocaleCode;
+  if (m_mode == Mode::Viewport)
```
Comment (ygorshenin) [inline search/processor.cpp]: Is it possible? Search is usually called when mode is not Viewport.

Comment (ygorshenin) [inline search/processor.cpp]: I mean, for viewport search there exists a SearchViewportPoints() function.

File: search/processor.cpp [hunk 3]
```diff
@@ -376,59 +335,51 @@ int Processor::GetCategoryLocales(int8_t(&arr)[3]) const
...
-  }
+  ::search::ForEachCategoryType(slice, arrLocales, localesCount, m_categories, forward<ToDo>(todo));
 }
 
 void Processor::Search(Results & results, size_t limit)
 {
-  if (m_tokens.empty())
-    SuggestStrings(results);
+  Geocoder::Params geocoderParams;
+  InitParams(geocoderParams);
```
Comment (ygorshenin) [inline search/processor.cpp]: I suggest you to create to aux functions:
```c++
InitGeocoderParams();
InitRankerParams();
```
and use them here. The code will be more declarative.

File: search/processor.cpp [hunk 4]
```diff
@@ -164,7 +125,8 @@ Processor::Processor(Index & index, CategoriesHolder const & categories,
       {StringUtf8Multilang::kInternationalCode, StringUtf8Multilang::kEnglishCode},
       {StringUtf8Multilang::kDefaultCode}};
 
-  m_keywordsScorer.SetLanguages(langPriorities);
+  m_ranker.m_keywordsScorer.SetLanguages(langPriorities);
+  m_preRanker.m_ranker = &m_ranker;
```
Comment (ygorshenin) [inline search/processor.cpp]: If you still need to create a reference to m_ranker in m_preRanker, why not to inverse dependency? From my point of view, m_preRanker should know about m_ranker, and call it on each PreResult1.

File: search/processor.cpp [hunk 5]
```diff
@@ -294,12 +253,12 @@ void Processor::SetRankPivot(m2::PointD const & pivot)
 
 void Processor::SetLanguage(int id, int8_t lang)
 {
-  m_keywordsScorer.SetLanguage(GetLangIndex(id), lang);
+  m_ranker.m_keywordsScorer.SetLanguage(GetLangIndex(id), lang);
```
Comment (ygorshenin) [inline search/processor.cpp]: May be it worth to protect keywords scorer, and create a bunch of public methods in Ranker instead?

File: search/processor.cpp (hunk 2 above)
Comment (ygorshenin) [inline search/processor.cpp]: On the other hand, we probably need to get rid of public SearchViewportPoints() and Search() functions. Currently SearchEngine dispatches calls to Processor. Probably we should do this dispatching inside Processor.

File: search/processor.hpp [hunk 6]
```diff
@@ -137,15 +130,10 @@ class Processor : public my::Cancellable
   friend class PreResult2Maker;
   friend class Ranker;
 
-  int GetCategoryLocales(int8_t(&arr)[3]) const;
-
-  void ForEachCategoryType(
-      StringSliceBase const & slice,
-      function<void(size_t /* tokenId */, uint32_t /* typeId */)> const & fn) const;
+  size_t GetCategoryLocales(int8_t(&arr)[3]) const;
```
Comment (ygorshenin) [inline search/processor.hpp]: Is clang-format happy with that?

File: search/processor.hpp [hunk 7]
```diff
@@ -157,34 +145,15 @@ class Processor : public my::Cancellable
...
-  void GetBestMatchName(FeatureType const & f, string & name) const;
-
-  Result MakeResult(PreResult2 const & r) const;
-  void MakeResultHighlight(Result & res) const;
-
-  Index & m_index;
   CategoriesHolder const & m_categories;
-  vector<Suggest> const & m_suggests;
   storage::CountryInfoGetter const & m_infoGetter;
+  TOnResults m_onResults;
```
Comment (ygorshenin) [inline search/processor.hpp]: Why do you need this?

File: search/ranker.cpp [hunk 8]
```diff
@@ -46,12 +48,93 @@ void RemoveDuplicatingLinear(vector<IndexedValue> & values)
...
+  using TIter = utf8::unchecked::iterator<string::const_iterator>;
+  TIter iter(str.end());
+  while (iter.base() != str.begin())
+  {
+    TIter prev = iter;
+    --prev;
+
+    if (delims(*prev))
+      break;
+    else
```
Comment (ygorshenin) [inline search/ranker.cpp]: No need to else after break.

File: search/ranker.cpp [hunk 9]
```diff
@@ -46,12 +48,93 @@ void RemoveDuplicatingLinear(vector<IndexedValue> & values)
                       }),
                values.end());
 }
+
+void RemoveStringPrefix(string const & str, string & res)
```
Comment (ygorshenin) [inline search/ranker.cpp]: The name is misleading. I guess it should be renamed to GetStringPrefix().

Comment (ygorshenin) [inline search/ranker.cpp]: With a comment what it does.

File: search/ranker.cpp [hunk 10]
```diff
@@ -46,12 +48,93 @@ void RemoveDuplicatingLinear(vector<IndexedValue> & values)
...
+  case STATE:
+  case CITY: return type;
+  case TOWN: return CITY;
+  case VILLAGE: return NONE;
+  case LOCALITY_COUNT: return type;
+  }
+}
+
+/// Makes continuous range for tokens and prefix.
+template <class TIter, class ValueT>
```
Comment (ygorshenin) [inline search/ranker.cpp]: s/ValueT/TValue/g

File: search/ranker.cpp [hunk 11]
```diff
@@ -46,12 +48,93 @@ void RemoveDuplicatingLinear(vector<IndexedValue> & values)
...
+class CombinedIter
+{
+  TIter m_i, m_end;
+  ValueT const * m_val;
+
+public:
+  CombinedIter(TIter i, TIter end, ValueT const * val) : m_i(i), m_end(end), m_val(val) {}
```
Comment (ygorshenin) [inline search/ranker.cpp]: s/i/cur/g

File: search/ranker.cpp [hunk 12]
```diff
@@ -46,12 +48,93 @@ void RemoveDuplicatingLinear(vector<IndexedValue> & values)
...
+  case VILLAGE: return NONE;
+  case LOCALITY_COUNT: return type;
+  }
+}
+
+/// Makes continuous range for tokens and prefix.
+template <class TIter, class ValueT>
+class CombinedIter
+{
+  TIter m_i, m_end;
```
Comment (ygorshenin) [inline search/ranker.cpp]: s/m_i/m_cur/g

Comment (ygorshenin) [inline search/ranker.cpp]: Please, separate these declarations.

File: search/ranker.cpp [hunk 13]
```diff
@@ -46,12 +48,93 @@ void RemoveDuplicatingLinear(vector<IndexedValue> & values)
...
+
+    return *m_val;
+  }
+
+  CombinedIter & operator++()
+  {
+    if (m_i != m_end)
+      ++m_i;
+    else
+      m_val = 0;
```
Comment (ygorshenin) [inline search/ranker.cpp]: The logic here is quite strange - m_val is set to zero when we call ++ on already finished iterator. Do we really need this?

File: search/ranker.cpp [hunk 14]
```diff
@@ -46,12 +48,93 @@ void RemoveDuplicatingLinear(vector<IndexedValue> & values)
...
+  case CITY: return type;
+  case TOWN: return CITY;
+  case VILLAGE: return NONE;
+  case LOCALITY_COUNT: return type;
+  }
+}
+
+/// Makes continuous range for tokens and prefix.
+template <class TIter, class ValueT>
+class CombinedIter
```
Comment (ygorshenin) [inline search/ranker.cpp]: Is it possible to get rid of CombinedIter (as it's used in a quite specific case) and use StringSlice instead?

File: search/ranker.cpp [hunk 15]
```diff
@@ -214,6 +306,160 @@ void Ranker::MakePreResult2(Geocoder::Params const & params, vector<IndexedValue
...
+    suggest.push_back(' ');
+  }
+}
+
+void Ranker::SuggestStrings(Results & res)
+{
+  if (m_params.m_prefix.empty() || !m_params.m_suggestsEnabled)
+    return;
+
+  string prolog;
```
Comment (ygorshenin) [inline search/ranker.cpp]: prologue?

File: search/ranker.hpp [hunk 16]
```diff
@@ -1,43 +1,127 @@
...
+    set<uint32_t> m_preferredTypes;
+    bool m_suggestsEnabled = false;
+
+    string m_query;
+    buffer_vector<strings::UniString, 32> m_tokens;
+    // Prefix of the last token in the query.
+    // We need it here to make suggestions.
+    strings::UniString m_prefix;
+
+    int8_t m_categoryLocales[3];
```
Comment (ygorshenin) [inline search/ranker.hpp]: Please, add a default ctor with zero initialization o this field.

File: search/ranker.hpp [hunk 17]
```diff
@@ -1,43 +1,127 @@
...
 
-  void FlushResults(Geocoder::Params const & params, Results & res, size_t resCount);
-  void FlushViewportResults(Geocoder::Params const & params, Results & res);
+  void SetLocalityFinderLanguage(int8_t code);
+
+  inline void BailIfCancelled() { ::search::BailIfCancelled(m_cancellable); }
+
+  KeywordLangMatcher m_keywordsScorer;
+
+  friend class PreResult2Maker;
```
Comment (ygorshenin) [inline search/ranker.hpp]: Feel free to move this to the private section.

File: search/ranker.hpp [hunk 18]
```diff
@@ -1,43 +1,127 @@
...
+
+  void ClearCaches();
 
-  void FlushResults(Geocoder::Params const & params, Results & res, size_t resCount);
-  void FlushViewportResults(Geocoder::Params const & params, Results & res);
+  void SetLocalityFinderLanguage(int8_t code);
+
+  inline void BailIfCancelled() { ::search::BailIfCancelled(m_cancellable); }
+
+  KeywordLangMatcher m_keywordsScorer;
```
Comment (ygorshenin) [inline search/ranker.hpp]: Is it possible to move this to the private section?

File: search/pre_ranker.hpp (hunk 1 above)
Comment (mpimenov) [inline search/pre_ranker.hpp]: Not in this manner.
done

File: search/processor.cpp (hunk 4 above)
Comment (mpimenov) [inline search/processor.cpp]: This code is left over after an experiment.
done

File: search/processor.cpp (hunk 5 above)
Comment (mpimenov) [inline search/processor.cpp]: done

File: search/processor.cpp (hunk 4 above)
Comment (mpimenov) [inline search/processor.cpp]: Oh. It is not that simple, but I'll think about it.

File: search/processor.cpp (hunk 3 above)
Comment (mpimenov) [inline search/processor.cpp]: done

File: search/processor.cpp (hunk 2 above)
Comment (mpimenov) [inline search/processor.cpp]: Ranker's use of m_viewport is guarded by the Mode::VIewport check.
Moving the search logic from Engine to Processor sounds good.

Comment (mpimenov) [inline search/processor.cpp]: Ranker needs viewport for an obscure normalisation of score. We should remove it later.

File: search/processor.hpp (hunk 6 above)
Comment (mpimenov) [inline search/processor.hpp]: Yes, but on second thought, buffer_vector is better here.

File: search/processor.hpp (hunk 7 above)
Comment (mpimenov) [inline search/processor.hpp]: I'll probably need it later. Removed for now.

File: search/ranker.cpp (hunk 9 above)
Comment (mpimenov) [inline search/ranker.cpp]: done

File: search/ranker.cpp (hunk 10 above)
Comment (mpimenov) [inline search/ranker.cpp]: done

File: search/ranker.cpp (hunk 12 above)
Comment (mpimenov) [inline search/ranker.cpp]: done

File: search/ranker.cpp (hunk 11 above)
Comment (mpimenov) [inline search/ranker.cpp]: done

File: search/ranker.cpp (hunk 15 above)
Comment (mpimenov) [inline search/ranker.cpp]: done

File: search/ranker.hpp (hunk 18 above)
Comment (mpimenov) [inline search/ranker.hpp]: done

File: search/ranker.hpp (hunk 17 above)
Comment (mpimenov) [inline search/ranker.hpp]: done

File: search/ranker.cpp (hunk 13 above)
Comment (mpimenov) [inline search/ranker.cpp]: Do we really want to touch this code?

Comment (ygorshenin) [inline search/ranker.cpp]: We don't, and can leave this as is, because "it works".

**Comment by therearesomewhocallmetim** [comment]
FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

**Comment by therearesomewhocallmetim** [comment]
JTTP FORMAT

File: search/ranker.cpp (hunk 8 above)
Comment (mpimenov) [inline search/ranker.cpp]: done

File: search/ranker.hpp (hunk 16 above)
Comment (mpimenov) [inline search/ranker.hpp]: Changed to buffer_vector.
done

File: search/ranker.cpp (hunk 13 above)
Comment (mpimenov) [inline search/ranker.cpp]: The highlighter code is more complicated than it should be but let's leave it as is for now.

Comment (ygorshenin) [inline search/ranker.cpp]: np

File: search/processor.cpp (hunk 4 above)
Comment (mpimenov) [inline search/processor.cpp]: This reference is removed for now. The dependency is to be resolved in another PR.

**Comment by mpimenov** [comment]
PTAL

File: search/pre_ranker.hpp [hunk 19]
```diff
@@ -2,6 +2,7 @@
 
 #include "search/intermediate_result.hpp"
 
+#include "base/logging.hpp"
```
Comment (ygorshenin) [inline search/pre_ranker.hpp]: Do we need this?

File: search/ranker.hpp [hunk 20]
```diff
@@ -1,43 +1,143 @@
...
 
+#define FIND_LOCALITY_TEST
+
+#ifdef FIND_LOCALITY_TEST
+#include "search/locality_finder.hpp"
+#endif  // FIND_LOCALITY_TEST
+
+class CategoriesHolder;
+class Index;
+namespace storage
```
Comment (ygorshenin) [inline search/ranker.hpp]: Please, insert a blank line here.

File: search/ranker.hpp [hunk 21]
```diff
@@ -1,43 +1,143 @@
...
+  void FlushResults(Geocoder::Params const & geocoderParams, Results & res, size_t resCount);
+  void FlushViewportResults(Geocoder::Params const & geocoderParams, Results & res);
+
+  void SetParams(Params const & params) { m_params = params; }
+
+  void ClearCaches();
+
+  void SetLocalityFinderLanguage(int8_t code);
+
+  void SetLanguages(vector<vector<int8_t>> const & languagePriorities)
```
Comment (ygorshenin) [inline search/ranker.hpp]: inline

File: search/ranker.hpp [hunk 22]
```diff
@@ -1,43 +1,143 @@
...
+  {
+    m_keywordsScorer.SetLanguages(languagePriorities);
+  }
+  void SetLanguage(pair<int, int> const & ind, int8_t lang)
```
Comment (ygorshenin) [inline search/ranker.hpp]: inline
Also, please, insert a blank line above.

File: search/pre_ranker.hpp (hunk 19 above)
Comment (mpimenov) [inline search/pre_ranker.hpp]: No.
done

File: search/ranker.hpp (hunk 20 above)
Comment (mpimenov) [inline search/ranker.hpp]: done

**Comment by therearesomewhocallmetim** [comment]
FORMAT JTTP

File: search/ranker.hpp (hunk 21 above)
Comment (mpimenov) [inline search/ranker.hpp]: done

**Comment by mpimenov** [comment]
PTAL

**Comment by ygorshenin** [comment]
LGTM
